src/pulsar_client_manager.erl

```diff
@@ -0,0 +1,381 @@
+%% Copyright (c) 2013-2019 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+
+-module(pulsar_client_manager).
+
+-feature(maybe_expr, enable).
+
+-behaviour(gen_server).
+
+-export([start_link/3]).
+
+%% gen_server Callbacks
+-export([ init/1
+        , handle_call/3
+        , handle_cast/2
+        , handle_info/2
+        , terminate/2
+        ]).
+
+-export([ get_topic_metadata/3
+        , lookup_topic/3
+        , lookup_topic_async/2
+        ]).
+
+-export([ get_status/2
+        , get_alive_pulsar_url/2
+        , get_workers/2
+        ]).
+
+-export_type([topic/0]).
+
+%%--------------------------------------------------------------------
+%% Type definitions
+%%--------------------------------------------------------------------
+
+-type client_id() :: atom().
+-type url() :: string().
+-type topic() :: string().
+-type partition_topic() :: string().
+
+-define(client_id, client_id).
+-define(initial_opts, initial_opts).
+-define(seed_urls, seed_urls).
+-define(workers, workers).
+-define(no_servers_available, no_servers_available).
+
+-type state() :: #{
+    ?client_id := client_id(),
+    ?initial_opts := map(),
+    ?seed_urls := [url()],
+    ?workers := #{url() => pid(), pid() => url()}
+}.
+
+%% calls/casts/infos
+-record(get_topic_metadata, {topic :: topic()}).
+-record(lookup_topic, {deadline :: infinity | integer(), partition_topic :: partition_topic()}).
+-record(lookup_topic_async, {from, partition_topic :: partition_topic()}).
+-record(get_status, {}).
+-record(get_alive_pulsar_url, {}).
+-record(get_workers, {}).
+
+%%--------------------------------------------------------------------
+%% API
+%%--------------------------------------------------------------------
+
+start_link(ClientId, Servers, Opts) ->
+    Args = #{
+      client_id => ClientId,
+      seed_urls => Servers,
+      initial_opts => Opts
+    },
+    gen_server:start_link({local, ClientId}, ?MODULE, Args, []).
+
+get_topic_metadata(ClientId, Topic, Timeout) ->
+    gen_server:call(ClientId, #get_topic_metadata{topic = Topic}, Timeout).
+
+lookup_topic(ClientId, PartitionTopic, Timeout) ->
+    Deadline = case is_integer(Timeout) of
+        true when Timeout > 0 -> now_ts() + Timeout;
+        false -> infinity
+    end,
+    gen_server:call(
+      ClientId,
+      #lookup_topic{deadline = Deadline, partition_topic = PartitionTopic},
```

Comment:
since other gen-calls use timeout, but this one uses deadline, 
maybe add a comment to explain why this gen-call has to use deadline

Reply:
:checkered_flag: